### ISSUE TITLE
deno deploy: Code generation from strings disallowed for this context

### ISSUE THREAD

**Issue by ci7lus** [b-fuze/deno-dom#75]
I tried to use this library in [deno deploy](https://deno.com/deploy), but I got the following error and it did not work.
```
Error serving request: EvalError: Code generation from strings disallowed for this context
    at Function (<anonymous>)
    at compile (https://deno.land/x/deno_dom@v0.1.19-alpha/src/dom/nwsapi.js:766:17)
    at collect (https://deno.land/x/deno_dom@v0.1.19-alpha/src/dom/nwsapi.js:1556:22)
    at _querySelectorAll (https://deno.land/x/deno_dom@v0.1.19-alpha/src/dom/nwsapi.js:1516:36)
    at Object._querySelector [as first] (https://deno.land/x/deno_dom@v0.1.19-alpha/src/dom/nwsapi.js:1422:14)
    at HTMLDocument.querySelector (https://deno.land/x/deno_dom@v0.1.19-alpha/src/dom/document.ts:160:28)
```
deno deploy does not seem to be able to execute functions dynamically. Are there any plans to support this?
Thanks.

**Comment by sno2** [comment]
> deno deploy does not seem to be able to execute functions dynamically. Are there any plans to support this? Thanks.

The problem is actually upstream as we use a static clone of [`nwsapi`](https://github.com/dperini/nwsapi) to parse query selectors.  I have created [an issue in their repo](https://github.com/dperini/nwsapi/issues/51) documenting this blocker for Deno Deploy support for Deno and will try and fix it later today when I get home.

**Comment by sno2** [comment]
You can read [my update](https://github.com/dperini/nwsapi/issues/51#issuecomment-986155145) on my (WIP) solution.

**Comment by b-fuze** [comment]
@sno2 did you consider using any other parser libraries out there?

**Comment by sno2** [comment]
> @sno2 did you consider using any other parser libraries out there?

Yes, although I couldn't find any that were maintained.  If I created my own, then I could maintain it as it's fairly simple TypeScript code.

**Comment by b-fuze** [comment]
How about [SizzleJS](https://github.com/jquery/sizzle)

**Comment by b-fuze** [comment]
It'd be best if the selector parser was as compliant with the spec as possible.

**Comment by sno2** [comment]
I did not see that in my NPM searches.  Although it seems to not be maintained and they also have a different goal of [not implementing features that can be added via JQuery](https://github.com/jquery/sizzle/pull/413#issuecomment-433223934).  Furthermore, it's written in JavaScript (not TypeScript) so I can't really work on a thing that large without the safety of TypeScript all in one file to make it fully spec-compatible.

**Comment by b-fuze** [comment]
Typescript does not ensure spec-compliance, unit tests do. What would be nice is something that is at least compliant with the majority of CSS2 selectors.

**Comment by sno2** [comment]
> Typescript does not ensure spec-compliance, unit tests do.

Yep, I was going to add the WPT tests which Sizzle does not use.

> What would be nice is something that is at least compliant with the majority of CSS2 selectors.

Yep, but CSS 3 has been standardized and not supported it would be a great mistake.

**Comment by b-fuze** [comment]
> Yep, but CSS 3 has been standardized and not supporting it would be a great mistake.

I thought you weren't implementing a fully (modern) spec-compliant selector parser. If you are then we can take a look at it.

**Comment by sno2** [comment]
> 

Oh, yeah I would make it fully spec-compatible.

**Comment by b-fuze** [comment]
@ci7lus I forgot to clarify that `document.getElementById/getElementsByClassName/getElementsByTagName` all _do_ work in Deno Deploy. As do their `Element.*` counterparts. Until we settle on a selector parser solution, you may find those helpful in the time being.

**Comment by BrunoBernardino** [comment]
I just ran into this problem and solved it with another external dependency, and it worked with deno deploy: [cheerio](https://github.com/Bestulo/deno-cheerio), in case it helps someone else.

Would very much prefer something smaller and closer to web standards, though.

**Comment by b-fuze** [comment]
This is fixed in `v0.1.22-alpha`. Deno DOM will detect if code generation is allowed, and if it isn't, it'll switch to the alternative selector engine.
